### PR TITLE
introduce MultiBsplineEvalHelper.hpp

### PR DESCRIPTION
**Pull request by ye-luo**
introduce MultiBsplineEvalHelper.hpp

All the helper functions are moved to this class. getSplineBound is moved there. computeLocationAndFractional is introduced to determine the location of an electron on the spline grid and compute all the prefactor. Separating this from the spline evaluation adds more flexibility. We will be able to manipulate location and spline coefficients storage easily.

In addition, the OpenMP offload code benefits from this change by avoid repeating this computation on every GPU thread and it results in register saving.

PS: Add two new allocator CUDALockedPageAllocator and PinnedAllocator/PinnedAlignedAllocator for precise host allocation control.

### CODE REVIEW

**Comment by ye-luo** [comment]
> Naming changes in the allocators could help with clarity.
> 

Is this just a comment or you need me to update some names?

> As a general issue the unit tests in spline2 are insufficient. The coverage needs to be better and if it focused more on individual calls it would be easier to see if corner cases were covered.
> 

The tests under spline2 don't cover the offload ones. I need some effort to have that coverage. A test  covers 3 cases, call these functions without offload, with offload and with offload but disabled.

> I'm just not sure the tests_splines<double, 128, 10> test covers that much of this code for CPU or OFFLOAD.
> 
> Are the cases of an offload spline eval kernel that gets all the splines versus when it gets split up via partial first and last covered? Are the thread block dimensions or thread block grid dimensions implied?
> 
splines get split up. Only the first to last section is evaluated. This is intended to explicitly use hierarchical parallelism. This is not CUDA and thread block dimensions or thread block grid dimensions
are not implied.

> I may just be missing how some of this is tested. Without some more tests I'm not completely sure how to review the changes in SplineC2ROMP.h.

We actually lack a diamondC_2x1x1 test to cover SplineC2R in general.
See test_einset.cpp

File: src/QMCWaveFunctions/BsplineFactory/SplineC2ROMP.h [hunk 1]
```diff
@@ -504,18 +493,23 @@ struct SplineC2ROMP : public SplineAdoptorBase<ST, 3>
...
 
       PRAGMA_OFFLOAD("omp target teams distribute num_teams(NumTeams) thread_limit(ChunkSizePerTeam) \
                   map(always, from: psi_ptr[0:orb_size])")
       for (int team_id = 0; team_id < NumTeams; team_id++)
       {
         const int first = ChunkSizePerTeam * team_id;
         const int last  = (first + ChunkSizePerTeam) > padded_size ? padded_size : first + ChunkSizePerTeam;
+
+        int ix, iy, iz;
+        ST a[4], b[4], c[4];
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2ROMP.h]: Here. I feel the current way is better than
```
int ix;
int iy;
int iz;
```

File: src/spline2/MultiBsplineEvalHelper.hpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * header file for helper functions
+ */
+#ifndef SPLINE2_MULTIEINSPLINE_EVAL_HELPER_HPP
+#define SPLINE2_MULTIEINSPLINE_EVAL_HELPER_HPP
+
+#include <cmath>
+#include <algorithm>
+
+namespace spline2
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/spline2/MultiBsplineEvalHelper.hpp]: I feel a bit wired that the first letter is lower case but the Bspline is in capital.
Is MultiBspline_helper.hpp better?

File: src/CUDA/CUDAallocator.hpp [hunk 3]
```diff
@@ -154,6 +154,41 @@ bool operator!=(const CUDAHostAllocator<T1>&, const CUDAHostAllocator<T2>&)
...
+    return pt;
+  }
+
+  void deallocate(value_type* pt, std::size_t n)
+  {
+    cudaErrorCheck(cudaHostUnregister(pt), "cudaHostUnregister failed in CUDALockedPageAllocator!");
+    HostAllocator::deallocate(pt, n);
+  }
+};
+
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/CUDA/CUDAallocator.hpp]: `use basic unit tests` could you be a bit more specific? Do you want me to add a simple unit test to cover this?

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/CUDA/CUDAallocator.hpp]: I think it would be useful to have the allocators tested through an  allocation, register, (device transfer), deregister, deallocate cycle.

File: src/CUDA/CUDAallocator.hpp [hunk 4]
```diff
@@ -154,6 +154,41 @@ bool operator!=(const CUDAHostAllocator<T1>&, const CUDAHostAllocator<T2>&)
 {
   return false;
 }
+
+/// allocator locks memory pages allocated by HostAllocator
+template<typename T, class HostAllocator = std::allocator<T>>
```
Comment (PDoakORNL) [inline src/CUDA/CUDAallocator.hpp]: Here is where I think the naming is an issue. Somehow I lost this comment so my final review didn't make much sense.
the Template parameter HostAllocator seems like it has a connection to the CudaHostAllocator  immediately above, but this is in fact an allocator that wouldn't be used with CUDALockedPageAllocator.  Changing the parameter to say `ULHA` (I like short all caps template param names) and defining that as an unlocked host allocator was what I was thinking.

File: src/spline2/MultiBsplineEvalHelper.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/spline2/MultiBsplineEvalHelper.hpp]: my thinks is multiBsplineHelperFunctions follows the function naming convention where the first letter is lcase, still Functions is redundant then. Do what you think is best for this PR, we can make function file naming consistent later.

**Comment by PDoakORNL** [comment]
So the tests will be in a future PR and you'll make sure they cover the relevant cases, which can be understood as hierarchical parallelism.

File: src/spline2/MultiBsplineEvalHelper.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/spline2/MultiBsplineEvalHelper.hpp]: Changed to MultiBsplineEval_helper.hpp

File: src/CUDA/CUDAallocator.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/CUDA/CUDAallocator.hpp]: Added unit tests to cover allocation, register, deregister, deallocate cycle. No transfer since this is not part of the allocator.

File: src/CUDA/CUDAallocator.hpp (hunk 4 above)
Comment (ye-luo) [reply src/CUDA/CUDAallocator.hpp]: Renamed template argument and add more documentation.
